File: packages/case-definition-dsl/src/matchers/http/HttpStatusCode/HttpStatusCode.ts
import { HTTP_STATUS_CODE_MATCHER_TYPE } from '@contract-case/case-core-plugin-http-dsl';
import { AnyMatcherWithExample } from '../../base';
import { codesToExample } from './codeToExample';

/**
 * Matches an HTTP status code
 * @public
 */
export class HttpStatusCode extends AnyMatcherWithExample {
  /** @internal */
  readonly '_case:matcher:type': typeof HTTP_STATUS_CODE_MATCHER_TYPE;

  /** @internal */
  readonly '_case:matcher:example': number;

  /** @internal */
  readonly '_case:matcher:rule': string | string[];

  /** @internal */
  readonly '_case:matcher:resolvesTo': 'HttpStatusCode';

  /**
   * @param statusCode - The http status code to match, expressed as a number or string (eg `200`, `"404"` or `"4XX"`).
   *
   * Matching can be relaxed with `X`, eg `"4XX"` or `"5XX"`. This is useful for error handling.
   *
   * If you need to match multiple specific statues, you can provide an array of string or numbers.
   * This behaviour is provided so that you can offer flexibility to the
   * implementation if the code path for multiple status codes is exactly the
   * same. It is not appropriate to use multiple status codes if the code paths
   * are intended to be different. For more context, see [the section on optional values
   * in the documentation](https://case.contract-testing.io/docs/faq#how-do-i-tell-contractcase-that-a-field-is-optional) for more details.
   */
  constructor(statusCode: string | string[]) {
    super(HTTP_STATUS_CODE_MATCHER_TYPE, codesToExample(statusCode));
    this['_case:matcher:rule'] = Array.isArray(statusCode)
      ? statusCode.map((r) => `${r}`)
      : `${statusCode}`;
  }

  /**
   * For non-TypeScript implementations (see `AnyMatcher.toJSON`)
   *
   * @privateRemarks
   * This comment and the implementation is boilerplate on all matchers to avoid
   * outputting duplicate unimportant documentation on all matcher classes of
   * the docs. Only modify this comment or the implementation via search and replace.
   */
  override toJSON(): unknown {
    return super.toJSON();
  }
}

File: packages/dsl-java/src/main/java/io/contract_testing/contractcase/dsl/matchers/http/HttpStatusCode.java
package io.contract_testing.contractcase.dsl.matchers.http;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.contract_testing.contractcase.dsl.ContractCaseDsl;
import io.contract_testing.contractcase.dsl.DslMatcher;
import jakarta.annotation.Generated;
import java.lang.Object;
import java.lang.String;
import lombok.Builder;
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

/**
 * Matches an HTTP status code.<p>This takes a string, so that you can relax matching,
 * with {@code X}, eg {@code "4XX"} or {@code "5XX"}. This is useful for covering error handling in contracts.<p>If you need to match multiple specific statues, you can use the HttpStatusCodes matcher instead, which takes an array.
 */
@Generated("@contract-case/case-definition-generator")
@ContractCaseDsl
public class HttpStatusCode implements DslMatcher {

  /**
   * ContractCase's internal type for this element
   */
  @Getter
  @JsonProperty("_case:matcher:type")
  private final String type;

  /**
   * Constant parameter resolvesTo
   */
  @Getter
  @JsonProperty("_case:matcher:resolvesTo")
  private final String resolvesTo = "string";

  /**
   * The status code to match
   */
  @Getter
  @JsonProperty("_case:matcher:rule")
  private final Object statusCode;

  /**
   * Matches an HTTP status code.<p>This takes a string, so that you can relax matching,
   * with {@code X}, eg {@code "4XX"} or {@code "5XX"}. This is useful for covering error handling in contracts.<p>If you need to match multiple specific statues, you can use the HttpStatusCodes matcher instead, which takes an array.
   * @param statusCode The status code to match
   */
  @Builder
  public HttpStatusCode(@NotNull final Object statusCode) {
    this.type = "_case:HttpStatusCode";
    this.statusCode = statusCode;
  }
}
